sessions/2018/october/Session_15.md

```diff
@@ -0,0 +1,324 @@
+
+- API using some backend programming language such as Java for example.
+- SPA (Single Page Application) is the frontend in which we can use JS, CSS, HTML5 or some JS framework such as React, Angular or Vue since the're some of the most popular SPA options right now.
+
+
+#### Dependency injection concepts
+
+A simplification and better practice to use dependency injection in Spring was talked about. We were changing from field injection towards constructor injection.
+
+##### Field injection
+
+Field injection allows to link resources, so for example in the next code snippet, a controller is asking to Spring to `@Autowire` the `BookService` class from its pool of registered services:
+
+```java
+import org.springframework.beans.factory.annotation.Autowired;
+
+@RestController
+@RequestMapping("/api/book")
+public class BookController {
+
+    @Autowired
+    private BookService bookService;
+    // ...
+}
+```
+
+__NOTE__: Field injection is the best choise for JUnit tests.
+
+
+##### Constructor injection
+
+Other way to use our service into our controller is to create a constructor that receives the service and populate it into a local field.
+
+So, once we create the constructor, we can declare `BookService` field as final (by using `@Autowired` that isn't possible). And even, we don't need anymore to use `@Autowired` annotation at all as is showed following:
+
+```java
+@RestController
+@RequestMapping("/api/book")
+public class BookController {
+
+    private final BookService bookService;
+
+    BookController(BookService bookService) {
+        this.bookService = bookService;
+    }
+
+    // ...
+}
+```
+
+__NOTE__: Constructor injection is the best choice to use in main code.
+
+__NOTE__: When we declare a field as final, it cannot be modified after, thus we have quicker applications and less prone to modifications by mistake in the fields we don't want. By other side, is not recommended use final keyword in classes and methods because that lead to applications that are hard to work with.
+
+#### Testing in Spring
+
+When we are testing services in Spring, we can built separate test files and if we are testing services that have not dependencies on Spring components we can use test in normal way without use `@Autowire` annotation (injection). For example:
+
+```java
+package co.org.osso.api;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+public class ApiApplicationBookTests {
+
+    private BookService BookServiceTarget = new BookService();
+
+    @Test
+    public void checkBookServiceGetBookWhenNull() {
+        Book book = BookServiceTarget.getBook(0);
+        Assert.assertEquals(null, book.getBookID());
+    }
+
+}
+```
+
+But if we have dependencies, for example __when testing a controller__, so we must to use `@Autowire` to link dependencies and call controller's methods directly, not by hitting the URLs of HTTP methods (GET, POST, PUT, etc.) in the controller maps. For example:
```

Comment:
Note that a better and more complete test would be to actually hit the HTTP endpoints themselves with JSON and retrieve the JSON back. Unfortunately, setting up and explaining this sort of test would take an entire session on its own, and there's so much else I can teach you instead.

In fact, it's this type of test (hitting HTTP endpoints with JSON) I tend to write instead of the `controller.myMethod(myPojo)` approach the most, as it gives me more confidence that I've set up everything correctly. If you're interested in seeing such a test, and want to spend time learning about this right away then remind me during our next session and I'll show you some tests of this form from one of my private repositories.